module/VuFind/src/VuFind/Config/PathResolver.php

```diff
@@ -0,0 +1,157 @@
+ * PHP version 7
+ *
+ * Copyright (C) Villanova University 2010.
+ * Copyright (C) The National Library of Finland 2022.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Config
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+namespace VuFind\Config;
+
+/**
+ * Configuration File Path Resolver
+ *
+ * @category VuFind
+ * @package  Config
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+class PathResolver
+{
+    /**
+     * Mode for getConfigPath: try to find a local file but fall back to base file
+     * if not available.
+     *
+     * @var int
+     */
+    public const MODE_AUTO = 0;
+
+    /**
+     * Mode for getConfigPath: try to find a local file.
+     *
+     * @var int
+     */
+    public const MODE_LOCAL = 1;
+
+    /**
+     * Mode for getConfigPath: get local config file path regardless of whether the
+     * file exists.
+     *
+     * @var int
+     */
+    public const MODE_LOCAL_FORCE = 2;
+
+    /**
+     * Mode for getConfigPath: get the base configuration file path.
+     *
+     * @var int
+     */
+    public const MODE_BASE = 3;
+
+    /**
+     * Default configuration path.
+     *
+     * @var string
+     */
+    public const DEFAULT_CONFIG_PATH = 'config/vufind';
+
+    /**
+     * Get the file path to the local configuration file (null if none found).
+     *
+     * @param string  $filename config file name
+     * @param ?string $path     path relative to VuFind base (optional; use null for
+     * default)
+     * @param bool    $force    force method to return path even if file does not
+     * exist (default = false, do not force)
+     *
+     * @return ?string
+     */
+    public function getLocalConfigPath(
+        string $filename,
+        ?string $path = null,
+        bool $force = false
+    ): ?string {
+        if (null === $path) {
+            $path = self::DEFAULT_CONFIG_PATH;
+        }
+        if (defined('LOCAL_OVERRIDE_DIR') && strlen(trim(LOCAL_OVERRIDE_DIR)) > 0) {
```

Comment:
How would you feel about eliminating internal reliance on constants like LOCAL_OVERRIDE_DIR and APPLICATION_PATH in favor of local properties, so we can move the constant processing into a factory and pass relevant values via constructor calls?

Reply:
Good idea, now done.